File: frontend/src/api/processes.ts
import http from './http'

export interface ProcessItem {
  process_id: string
  name: string
  channel?: string
  description?: string
  entrypoints?: string[]
}

export interface ProcessDetail extends ProcessItem {}

export interface ProcessStep {
  step_id: number
  process_id: string
  order_no: number
  name?: string
  capability_id?: string
}

export interface ProcessEdge {
  id: number
  from_step_id: string
  to_step_id: string
  edge_type?: string
  condition?: string
  label?: string
}

export async function listProcesses(): Promise<ProcessItem[]> {
  const res = await http.get<ProcessItem[]>('/processes/list_processes')
  return res.data
}

export async function getProcess(processId: string): Promise<ProcessDetail> {
  const res = await http.get<ProcessDetail>('/processes/get_process', {
    params: { process_id: processId },
  })
  return res.data
}

export async function createProcess(payload: ProcessDetail): Promise<ProcessDetail> {
  const res = await http.post<ProcessDetail>('/processes/create_process', payload)
  return res.data
}

export async function updateProcess(processId: string, payload: Partial<ProcessDetail>): Promise<ProcessDetail> {
  const res = await http.post<ProcessDetail>('/processes/update_process', {
    process_id: processId,
    ...payload,
  })
  return res.data
}

export async function deleteProcess(processId: string): Promise<void> {
  await http.post('/processes/delete_process', { process_id: processId })
}

export async function getProcessSteps(processId: string): Promise<ProcessStep[]> {
  const res = await http.get<ProcessStep[]>(`/processes/get_process_steps`, {
    params: { process_id: processId },
  })
  return res.data
}

export async function saveProcessSteps(processId: string, steps: ProcessStep[]): Promise<ProcessStep[]> {
  const res = await http.post<ProcessStep[]>(`/processes/save_process_steps`, {
    process_id: processId,
    steps,
  })
  return res.data
}

export async function deleteProcessStep(processId: string, stepId: number): Promise<void> {
  await http.post('/processes/delete_process_step', {
    process_id: processId,
    step_id: stepId,
  })
}

export async function listProcessEdges(processId: string): Promise<ProcessEdge[]> {
  const res = await http.get<ProcessEdge[]>('/processes/list_process_edges', {
    params: { process_id: processId },
  })
  return res.data
}

export async function createProcessEdge(
  processId: string,
  payload: Omit<ProcessEdge, 'id'>,
): Promise<ProcessEdge> {
  const res = await http.post<ProcessEdge>('/processes/create_process_edge', {
    process_id: processId,
    ...payload,
  })
  return res.data
}

export async function updateProcessEdge(
  processId: string,
  edgeId: number,
  payload: Partial<Omit<ProcessEdge, 'id'>>,
): Promise<ProcessEdge> {
  const res = await http.post<ProcessEdge>('/processes/update_process_edge', {
    process_id: processId,
    edge_id: edgeId,
    ...payload,
  })
  return res.data
}

export async function deleteProcessEdge(processId: string, edgeId: number): Promise<void> {
  await http.post('/processes/delete_process_edge', {
    process_id: processId,
    edge_id: edgeId,
  })
}

export interface PublishResult {
  success: boolean
  message: string
  synced_at?: string
  error_type?: string
  stats?: {
    steps: number
    implementations: number
    data_resources: number
  }
}

export async function publishProcess(processId: string): Promise<PublishResult> {
  const res = await http.post<PublishResult>('/processes/publish_process', {
    process_id: processId,
  })
  return res.data
}

File: backend/app/schemas/processes.py
from typing import List

from pydantic import BaseModel


class ProcessBase(BaseModel):
    name: str
    channel: str | None = None
    description: str | None = None
    entrypoints: List[str] | None = None


class ProcessCreate(ProcessBase):
    process_id: str


class ProcessUpdate(ProcessBase):
    pass


class ProcessStep(BaseModel):
    step_id: int
    process_id: str
    order_no: int
    name: str | None = None
    capability_id: str | None = None


class ProcessEdgeBase(BaseModel):
    from_step_id: str
    to_step_id: str
    edge_type: str | None = None
    condition: str | None = None
    label: str | None = None


class ProcessEdgeCreate(ProcessEdgeBase):
    pass


class ProcessEdgeUpdate(BaseModel):
    from_step_id: str | None = None
    to_step_id: str | None = None
    edge_type: str | None = None
    condition: str | None = None
    label: str | None = None


class ProcessEdgeOut(ProcessEdgeBase):
    id: int

    class Config:
        from_attributes = True


class ProcessIdRequest(BaseModel):
    """通用的流程 ID 请求体。"""

    process_id: str


class ProcessUpdatePayload(ProcessUpdate):
    """更新流程时使用的请求体，包含流程 ID 及可更新字段。"""

    process_id: str


class SaveProcessStepsRequest(BaseModel):
    """保存流程步骤列表的请求体。"""

    process_id: str
    steps: List[ProcessStep]


class DeleteProcessStepRequest(BaseModel):
    """删除单个流程步骤的请求体。"""

    process_id: str
    step_id: int


class CreateProcessEdgeRequest(ProcessEdgeCreate):
    """在流程中创建边的请求体。"""

    process_id: str


class UpdateProcessEdgeRequest(ProcessEdgeUpdate):
    """更新流程边属性的请求体。"""

    process_id: str
    edge_id: int


class DeleteProcessEdgeRequest(BaseModel):
    """删除流程边的请求体。"""

    process_id: str
    edge_id: int

